Shell/Shuffling.hpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+/**
+ * @file Shuffling.hpp
+ * Defines class Shuffling implementing the shuffling of the input.
+ * @since 9/6/2021 Prague
+ */
+
+#ifndef __Shuffling__
+#define __Shuffling__
+
+#include "Forwards.hpp"
+
+#include "Lib/List.hpp"
+#include "Lib/Coproduct.hpp"
+#include "Kernel/Unit.hpp"
+
+#include <algorithm>
+
+namespace Shell {
+
+using namespace Lib;
+using namespace Kernel;
+
+/**
+ * Class implementing shuffling-related procedures.
+ * @since 09/06/2021 Prague
+ */
+class Shuffling
+{
+private:
+  typedef Coproduct<Formula*, Literal*, TermList> Shufflable;
```

Comment:
Joe-code is gaining ground! I'm fine with it, but does this get us something versus e.g. overloads, template specialisation?

Reply:
If I remember correctly, I interpreted this is a modern (type safe-ish and better documenting) way of doing tagged union. For that use case, it was a very pleasant experience!

Reply:
Ahah, you want a stack of heterogenous things in `shuffleIter`, right. OK, sounds good!